openSearchLane.go

```diff
@@ -0,0 +1,305 @@
+	go osl.openSearchConnection.processConnection(osl)
+
+	return
+
+}
+
+func (osl *openSearchLane) Derive() Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(osl.Context, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	osl.openSearchConnection.refCount++
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	osl.openSearchConnection.refCount++
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) Metadata(key, value string) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	if osl.metadata == nil {
+		osl.metadata = make(map[string]string)
+		osl.metadata["timestamp"] = time.Now().String()
+	}
+
+	osl.metadata[key] = value
+
+}
+
+func (osl *openSearchLane) SetEmergencyHandler() {
+	//TODO
+	if len(osl.openSearchConnection.logBuffer) > 100 {
+		fmt.Println("Write to disk")
+	}
+}
+
+func (osl *openSearchLane) Close() {
+	osl.openSearchConnection.refCountCh <- &osl.wg
+	osl.wg.Wait()
+}
+
+func (osl *openSearchLane) refCountRoutine() {
+	for {
+		select {
+		case wg := <-osl.openSearchConnection.refCountCh:
+			osl.openSearchConnection.mu.Lock()
+
+			osl.openSearchConnection.refCount--
+
+			wg.Done()
+
+			if osl.openSearchConnection.refCount == 0 {
+				fmt.Println("hmmmmm")
+				err := osl.openSearchConnection.flush(osl, osl.openSearchConnection.logBuffer)
+				if err != nil {
+					fmt.Print("error inserting logs in openSearch on lane termination")
+				}
+				return
+			}
+
+			osl.openSearchConnection.mu.Unlock()
+		}
+	}
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &LogTemplate{
+		AppName:      osl.openSearchConnection.appName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+		Metadata:     osl.metadata,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	return len(p), nil
+}
+
+func (osc *openSearchConnection) processConnection(l Lane) {
+	var lb []*LogTemplate
+	backoffDuration := 10 * time.Second
+	ticker := time.NewTicker(backoffDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			
+			osc.mu.Lock()
+			lb = append(lb, osc.logBuffer...)
```

Comment:
Considering this implementation, maintaining a common buffer for the EmergencyWriter function might pose challenges, especially when lb is not equal to logBuffer when opensearch is offline. I recommend reverting this change and simply managing the single reference (logBuffer).